plugins/out_s3/s3.c

```diff
@@ -423,6 +485,67 @@ static int cb_s3_init(struct flb_output_instance *ins,
             flb_plg_error(ctx->ins, "'s3_key_format' must start with a '/'");
             return -1;
         }
+        if (strstr((char *) tmp, "$INDEX")) {
+            ctx->key_fmt_has_seq_index = FLB_TRUE;
+
+            /* Initialize local storage to store metadata files */
+            ret = s3_store_init(ctx);
+            if (ret == -1) {
+                flb_plg_error(ctx->ins, "Failed to initialize S3 storage: %s",
+                              ctx->store_dir);
+                return -1;
+            }
+
+            fs_stream = flb_fstore_stream_create(ctx->fs, "metadata");
+            if (!fs_stream) {
+                flb_plg_error(ctx->ins, "could not initialize metadata stream");
+                flb_fstore_destroy(ctx->fs);
+                ctx->fs = NULL;
+                return -1;
+            }
+
+            /* Construct directories and file path names */
+            tmp = fs_stream->path;
+            ctx->metadata_dir = flb_sds_create(tmp);
```

Comment:
for all of these calls to flb_sds_create and flb_sds_cat- they can fail to alloc in which case you need to check the return value and then call `flb_errno()` if it failed.